### ISSUE TITLE
[receiver/filelog] `poll_interval` is not respected when files are too large

### ISSUE THREAD

**Issue by hvaghani221** [open-telemetry/opentelemetry-collector-contrib#17846]
### Component(s)

receiver/filelog

### Describe the issue you're reporting

 We are using filelog receiver to read container logs In [Splunk OpenTelemetry Collector for Kubernetes](https://github.com/signalfx/splunk-otel-collector-chart).

One of our customers noticed that logs were missing for some of the apps. On further inspection, we found that one container was producing logs too quickly and it was rotated in few range of 105ms-6s. We observed that some log files were completely missing. The reason could be that file was rotated before it was picked by the agent.

By inspecting the file consumer from stanza, I found that poll is done synchronously.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/cdf0a428caf65c83a71561dec41c97d4dca5e40f/pkg/stanza/fileconsumer/file.go#L101 

So, the time interval between 2 polls is the max of poll_interval and the time taken by the previous poll. And `poll_interval` will not be respected when a single poll takes more time than `poll_interval`.

The `poll` method finds the path of all the files and consumes them in batches of length `MaxConcurrentFiles/2` to limit the number of opened file descriptors.

Here, until all the paths from the batch are read to the end, it will not process the next batch. And the next batch is waiting for the current batch to finish. So, most of the time, it will be reading files lesser than `MaxConcurrentFiles/2`.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/cdf0a428caf65c83a71561dec41c97d4dca5e40f/pkg/stanza/fileconsumer/file.go#L132-L140

I think we can move to a thread pooling pattern to utilise MaxConcurrentFiles.

**Comment by harshit-splunk** [comment]
cc: @atoulme

**Comment by matthewmodestino** [comment]
In practice, we have seen many Kubernetes providers using default kubelet file rotation of 10MB, which is not a production ready default in my experience.

This causes any real production service with any real traffic load to spin files at a ridiculous rate.

As part of this issue we should be championing better production defaults in the Kubernetes community, as anyone who has worked in Production services before Kubernetes knows, rotation is generally in the realm of 500MB to 1GB or higher!!

Do we know what flavour of Kubernetes this particular example involves? (Open Source K8s or a managed provider?)

**Comment by harshit-splunk** [comment]
> Do we know what flavour of Kubernetes this particular example involves? (Open Source K8s or a managed provider?)

They are using GKE. Unfortunately, GKE doesn't expose container-log-max-size config (https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config). So, they are stuck with the default value.

**Comment by matthewmodestino** [comment]
Yeah, we need to make some noise with cloud providers on this. I wonder if they are able to request through their support. Will reach out and see what I can find out!

**Comment by djaglowski** [comment]
> I think we can move to a thread pooling pattern to utilise MaxConcurrentFiles.

I agree there is an opportunity to increase utilization of the allowed resources here.

One consideration to keep in mind is that we are currently leaning on the batching strategy to detect files that are rotated out of the receiver's include pattern. A pool based solution may also require rethinking of the detection algorithm.

Given that this may potentially involve some major changes, I believe we should look at adding a parallel implementation that can be opted into with a feature gate. Once proven out sufficiently, we can likely migrate to this strategy as default.

**Comment by djaglowski** [comment]
@VihasMakwana, thanks for sharing these benchmarks. Looks like a nice improvement.

I'll comment on your PR as well but I think it will be important to have benchmarking for this scenario as well as other (e.g. all similar sized files) in our codebase so that we can ensure all changes move us in the right direction.

**Comment by VihasMakwana** [comment]
@djaglowski makes sense. 
I performed the same benchmarks for similar files. Results are similar for both the models:

100 files of 160KB each: takes ~600ms to receive all the logs.
100 files of 1.6MB each: takes ~2.7s to receive all the logs.
100 files of 17MB each: takes ~26-27s to receive all the logs

**Comment by VihasMakwana** [comment]
I will add those test cases in my PR.

**Comment by paulojmdias** [comment]
I've been working on the ability for pooling asyncronously under feature gate (in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45129). Enabling it makes the async dispatch in `startPoller` decouple the ticker from poll execution, so `poll_interval` is respected regardless of how long a poll takes.

Technically, the semaphore I added in `consume()` is redundant, as the batch is already sliced to `maxBatchFiles` before reaching `consume()`, so the semaphore never actually blocks. I'll remove it in the next iteration. 👍 

However, the pre-existing sequential batch processing (`wg.Wait()` between batches) still means slow readers in one batch block discovery of files in the next. A shared worker pool fed with all matched files would better utilize `MaxConcurrentFiles`IMO. @andrzej-stencel WDYT of updating `consume()` in that direction as part of this feature gate?

I would like to have your input here, even before marking the PR #45129 as ready for review, as IdK if this is something we really want to support 🙏
